### PR TITLE
Config with pure python

### PR DESCRIPTION
**Pull request by jugmac00**
Please have a close look.

I had to make some adjustments because there is no `src` top level directory, ie I had to change the flake8 call and also the coverage run.

Other than that.. woot - nice work!

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -1,7 +1,20 @@
...
 recursive-include zc *.txt
+recursive-include docs Makefile
 
-global-exclude *.pyc
-global-exclude *.pyo
+recursive-include src *.pt
+recursive-include src *.rst
+recursive-include src *.txt
+recursive-include src *.xml
+recursive-include src *.zcml
```
Comment (dataflake) [inline MANIFEST.in:20]: This should probably be changed to `zc` instead of `src`

Comment (dataflake) [reply MANIFEST.in:20]: And then remove the duplicate `recursive-include zc *.txt`

Comment (jugmac00) [reply MANIFEST.in]: Thank you!

**Comment by icemac** [comment]
Probably it would have been better to add a `src` directory and move `zc` there.
Now this package cannot be updates without doing all the changes again.

**Comment by dataflake** [comment]
A strong `-1` from me for forcing package structure changes just to bend a package to a tool that could add intelligence to work around that instead. There are tons of packages that do not use a `src` toplevel folder.

**Comment by jamadden** [comment]
Countered with a strong +1 from me :) Not having a `src/` directory is a dangerous leftover.

**Comment by dataflake** [comment]
I can't see any particular advantage in having yet another layer of subfolder added.

**Comment by jamadden** [comment]
Since `.` is automatically on `PYTHONPATH`, it's far too easy to wind up pulling in some random code that you didn't intend to if there's no `src/`. The resulting errors can be maddening (broken namespace packages are especially common); it's bit me more than once.

Aesthetically I personally find it far more symmetrically pleasing, as well, for the same reason buildout creates a `parts/` directory. There's a clean separation between metadata and code.

**Comment by dataflake** [comment]
I personally never had any issues with broken namespece packages or inadvertently pulling in wrong code on any of the packages that I usually work on, like "real" Zope product packages. Most do not have the extra folder.

**Comment by icemac** [comment]
I had problems with a missing `src` folder, too. The easiest solution was always to create one instead of trying to work around.

**Comment by mgedmin** [comment]
https://hynek.me/articles/testing-packaging/ is a good read for the advantages of using src/ layouts.

I too would disagree with "forcing package structure changes just to bend a package to a tool that could add intelligence to work around that instead", but src/ layouts have other advantages, and I think we want them because of those.

File: zc/zdaemonrecipe/tests.py [hunk 2]
```diff
@@ -66,7 +66,7 @@ def newlines_in_program():
         path /sample-buildout/parts/run/transcript.log
       </logfile>
     </eventlog>
-    """
+    """  # noqa: F821
```
Comment (mgedmin) [inline zc/zdaemonrecipe/tests.py:69]: I'm starting to understand @jamadden's aversion to these ;)

F821 is "undefined name", and I think we could instead tell flake8 about the special names that zc.buildout's test setup makes available to test.globs, by adding a 

```
[flake8]
builtins = cat, system, join, write  # and whatever
```

to setup.cfg.

Comment (jugmac00) [reply zc/zdaemonrecipe/tests.py:69]: see https://github.com/zopefoundation/zc.zdaemonrecipe/pull/5

**Comment by dataflake** [comment]
> https://hynek.me/articles/testing-packaging/ is a good read for the advantages of using src/ layouts.
> 
> I too would disagree with "forcing package structure changes just to bend a package to a tool that could add intelligence to work around that instead", but src/ layouts have other advantages, and I think we want them because of those.

I found the article really interesting and investigated a little more. It seems that even among the packages we have that do have a `src` folder there are probably very few (None?) that correctly handle what I read as the main point: Make sure your tests run over the installed version of the package, not just the source tree.

I looked at `DocumentTemplate` and experimented for a bit to find a combination that appears to work. Here's what else I needed to change:

- remove `develop = .` in `buildout.cfg`
- make sure I tell buildout to not look for a stable egg by specifying `versions = versions` under `[buildout]` and adding a `[versions]` stanza with `DocumentTemplate =`
- in the `[test]` buildout config add `defaults = ['-sDocumentTemplate']` to make sure the test runner only looks for DocumentTemplate tests. This restriction is needed because the generated `test` script looks in the tox environment `site-packages` folder, which may contain other stuff.
- change `tox.ini` to add a `[coverage:paths]` stanza and change `[coverage:run]` so `source` is now `DocumentTemplate` instead of `src` (this is described in https://hynek.me/articles/testing-packaging)
- in `tox.ini` remove any `skip_install = true`, which prevents the installation of the egg into the tox environments

Many of the options I needed to change are common in our packages. So making sure to "do the right thing" is a bit of work. Unless I did this the wrong way. Comments welcome.

**Comment by mgedmin** [comment]
I've found it easier to ignore zc.buildout and use tox with zope-testrunner.  Tox has usedevelop=false by default and therefore tests the installed package.

(The recent [meta config tool](https://github.com/zopefoundation/meta/blob/master/config/pure-python/tox.ini) sets up tox configurations with usedevelop=true.  I don't feel strongly enough about the whole issue to object.  I find check-manifest helps avoid most of the problems that testing installed does.)

**Comment by dataflake** [comment]
So at this point my takeaway is this: I believe I understand the argument made by Hynek and why it's a good idea. But trying this myself I noticed that converting an existing package to not just use a `src` subfolder, but to change all necessary configurations so it really does "the right thing" and tests installed packages as opposed to just the source tree is a lot of work when starting from the common configurations found all over our packages - including those that already use a `src` structure. I do not think any meta config tool would be able to handle all that. If you get a similar effect with just pulling in a simple tool like `check-manifest` then I don't see the benefit anymore.

**Pull request merged**